src/webgpu/api/validation/createTexture.spec.ts

```diff
@@ -71,58 +68,149 @@ class F extends ValidationTest {
+    { size: [32, 1, 1], dimension: '1d' as const, mipLevelCount: 6, _success: true }, // full mip chains for 1D texture are allowed (Mip level sizes: 32, 16, 8, 4, 2, 1)
+    { size: [32, 1, 1], dimension: '1d' as const, mipLevelCount: 7, _success: false }, // too big mip chains for 1D texture are not allowed (Mip level sizes: 32, 16, 8, 4, 2, 1)
+    { size: [31, 1, 1], dimension: '1d' as const, mipLevelCount: 5, _success: true }, // full mip chains for non-power-of-two 1D texture are allowed (Mip level sizes: 31, 15, 7, 3, 1)
+    { size: [31, 1, 1], dimension: '1d' as const, mipLevelCount: 6, _success: false }, // too big mip chains for non-power-of-two 1D texture are not allowed (Mip level sizes: 31, 15, 7, 3, 1)
+    { size: [32, 32, 32], dimension: '3d' as const, mipLevelCount: 6, _success: true }, // full mip chains for 3D texture are allowed (Mip level sizes: 32x32x32, 16x16x16, 8x8x8, 4x4x4, 2x2x2, 1x1x1)
+    { size: [32, 32, 32], dimension: '3d' as const, mipLevelCount: 7, _success: false }, // too big mip chains for 3D texture are not allowed (Mip level sizes: 32x32x32, 16x16x16, 8x8x8, 4x4x4, 2x2x2, 1x1x1)
+    { size: [31, 31, 31], dimension: '3d' as const, mipLevelCount: 5, _success: true }, // full mip chains for non-power-of-two 3D texture are allowed (Mip level sizes: 31x31x31, 15x15x15, 7x7x7, 3x3x3, 1x1x1)
+    { size: [31, 31, 31], dimension: '3d' as const, mipLevelCount: 6, _success: false }, // too big mip chains for non-power-of-two 3D texture are not allowed (Mip level sizes: 31x31x31, 15x15x15, 7x7x7, 3x3x3, 1x1x1)
+    { size: [32, 32], mipLevelCount: 100, _success: false }, // undefined shift check if miplevel is bigger than the integer bit width
+    { size: [32, 8], mipLevelCount: 6, _success: true }, // non square mip map halves the resolution until a 1x1 dimension. (Mip maps: 32x8, 16x4, 8x2, 4x1, 2x1, 1x1)
+    { size: [32, 8], mipLevelCount: 7, _success: true }, // too big mip chains for non square mip map halves are disallowed (Mip maps: 32x8, 16x4, 8x2, 4x1, 2x1, 1x1)
+    { size: [32, 32, 64], mipLevelCount: 7, _success: false }, // array layer count for 2D texture should not be taken account to calculate mip levels. (Mip maps: 32x32x64, 16x16x64, 8x8x64, 4x4x64, 2x2x64, 1x1x64)
     {
-      width: 32,
-      height: 32,
-      arrayLayerCount: 64,
-      dimension: '3d',
+      size: [32, 32, 64],
+      dimension: '3d' as const,
       mipLevelCount: 7,
       _success: true,
-    }, // depth of 3D texture should be taken account to calculate mip levels. (Mip maps: 32 * 32 * 64, 16 * 16 * 32, 8 * 8 * 16, 4 * 4 * 8, 2 * 2 * 4, 1 * 1 * 2, 1 * 1 * 1)
+    }, // depth of 3D texture should be taken account to calculate mip levels. (Mip maps: 32x32x64, 16x16x32, 8x8x16, 4x4x8, 2x2x4, 1x1x2, 1x1x1)
   ])
   .fn(async t => {
-    const { width, height, mipLevelCount, arrayLayerCount, _success } = t.params;
+    const { size, mipLevelCount, dimension, _success } = t.params;
```

Comment:
thought to consider: It would be much easier to add cases if some part of this were computed instead of written manually. How about computing the max mipLevelCount and checking that's valid and that+1 is invalid? Computing it isn't quite trivial but it's fairly simple. A helper for it could live in `webgpu/util/texture`.

This would require separating the mipLevelCount:100 case, but that's testing a different thing anyway so it should be separate.

Reply:
Done. Added a function named maxMipLevelCount in util/texture/base.ts (please propose a new file name if you think it is not good enough). MipLevelCount:100 should still fail even according to the calculation, should we remove this test, or put it in a standalone test category?

Reply:
It should be a separate test.

Reply:
Done